File: Framework/Emlid.WindowsIoT.Hardware/Components/I2cExtensions.cs
using System;
using Windows.Devices.I2c;

namespace Emlid.WindowsIot.Hardware.Components
{
    /// <summary>
    /// Extensions for work with I2C devices.
    /// </summary>
    [CLSCompliant(false)]
    public static class I2cExtensions
    {
        #region Read

        /// <summary>
        /// Writes data then reads a single byte result.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <returns>Read data byte.</returns>
        public static byte WriteReadByte(this I2cDevice device, byte writeData)
        {
            // Call overloaded method
            return WriteReadBytes(device, new[] { writeData }, 1)[0];
        }

        /// <summary>
        /// Writes data then reads a single byte result.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <returns>Read data byte.</returns>
        public static byte WriteReadByte(this I2cDevice device, byte[] writeData)
        {
            // Call overloaded method
            return WriteReadBytes(device, writeData, 1)[0];
        }

        /// <summary>
        /// Writes data then reads a single byte result.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <param name="size">Amount of data to read.</param>
        /// <returns>Read data bytes.</returns>
        public static byte[] WriteReadBytes(this I2cDevice device, byte writeData, int size)
        {
            // Call overloaded method
            return WriteReadBytes(device, new[] { writeData }, size);
        }

        /// <summary>
        /// Writes data then reads one or more bytes.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <param name="size">Amount of data to read.</param>
        /// <returns>Read data bytes.</returns>
        public static byte[] WriteReadBytes(this I2cDevice device, byte[] writeData, int size)
        {
            var buffer = new byte[size];
            device.WriteRead(writeData, buffer);
            return buffer;
        }

        /// <summary>
        /// Writes data then reads one or more bytes into an existing buffer.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <param name="size">Amount of data to read.</param>
        /// <param name="buffer">Target buffer.</param>
        /// <param name="offset">Target buffer offset.</param>
        public static void WriteReadBytes(this I2cDevice device, byte writeData, int size, byte[] buffer, int offset)
        {
            // Call overloaded method
            WriteReadBytes(device, new[] { writeData }, size, buffer, offset);
        }

        /// <summary>
        /// Writes data then reads one or more bytes into an existing buffer.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <param name="size">Amount of data to read.</param>
        /// <param name="buffer">Target buffer.</param>
        /// <param name="offset">Target buffer offset.</param>
        public static void WriteReadBytes(this I2cDevice device, byte[] writeData, int size, byte[] buffer, int offset)
        {
            // Call overloaded method
            var data = WriteReadBytes(device, writeData, size);

            // Copy data to target buffer
            Array.ConstrainedCopy(data, 0, buffer, offset, size);
        }

        /// <summary>
        /// Writes data, reads a byte result then tests on or more bits.
        /// </summary>
        /// <remarks>
        /// Commonly used to test register flags.
        /// </remarks>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <param name="mask">Index of the bit to read, zero based.</param>
        /// <returns>True when the result was positive (any bits in the mask were set).</returns>
        public static bool WriteReadBit(this I2cDevice device, byte writeData, byte mask)
        {
            // Call overloaded method
            return WriteReadBit(device, new[] { writeData }, mask);
        }

        /// <summary>
        /// Writes data, reads a byte result then tests on or more bits.
        /// </summary>
        /// <remarks>
        /// Commonly used to test register flags.
        /// </remarks>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <param name="mask">Index of the bit to read, zero based.</param>
        /// <returns>True when the result was positive (any bits in the mask were set).</returns>
        public static bool WriteReadBit(this I2cDevice device, byte[] writeData, byte mask)
        {
            // Read byte
            var value = WriteReadByte(device, writeData);

            // Apply mask and return true when set
            return (value & mask) != 0;
        }

        #endregion

        #region Write

        /// <summary>
        /// Joins two byte values then writes them.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="data1">First part of data to write.</param>
        /// <param name="data2">Second part of data to write.</param>
        public static void WriteJoinByte(this I2cDevice device, byte data1, byte data2)
        {
            device.Write(new[] { data1, data2 });
        }

        /// <summary>
        /// Joins two byte values then writes them.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="data1">First part of data to write.</param>
        /// <param name="data2">Second part of data to write.</param>
        public static void WriteJoinByte(this I2cDevice device, byte[] data1, byte data2)
        {
            // Call overloaded method
            WriteJoinBytes(device, data1, new[] { data2 });
        }

        /// <summary>
        /// Joins two byte values then writes them.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="data1">First part of data to write.</param>
        /// <param name="data2">Second part of data to write.</param>
        public static void WriteJoinBytes(this I2cDevice device, byte data1, byte[] data2)
        {
            // Call overloaded method
            WriteJoinBytes(device, new[] { data1 }, data2);
        }

        /// <summary>
        /// Joins two byte values then writes them.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="data1">First part of data to write.</param>
        /// <param name="data2">Second part of data to write.</param>
        public static void WriteJoinBytes(this I2cDevice device, byte[] data1, byte[] data2)
        {
            var addressLength = data1.Length;
            var dataLength = data2.Length;
            var buffer = new byte[addressLength + dataLength];
            Array.Copy(data1, buffer, addressLength);
            Array.ConstrainedCopy(data2, 0, buffer, addressLength, dataLength);
            device.Write(buffer);
        }

        /// <summary>
        /// Sets or clears one or more bits.
        /// </summary>
        /// <param name="device">Device to use.</param>
        /// <param name="writeData">Data to write.</param>
        /// <param name="mask">
        /// Mask of the bit to set or clear according to value.
        /// Supports setting or clearing multiple bits.
        /// </param>
        /// <param name="value">Value of the bits, i.e. set or clear.</param>
        /// <returns>Value written.</returns>
        /// <remarks>
        /// Commonly used to set register flags.
        /// Reads the current byte value, merges the positive or negative bit mask according to value,
        /// then writes the modified byte back.
        /// </remarks>
        public static byte WriteReadWriteBit(this I2cDevice device, byte writeData, byte mask, bool value)
        {
            // Read existing byte
            var oldByte = WriteReadByte(device, writeData);

            // Merge bit (set or clear bit accordingly)
            var newByte = value ? (byte)(oldByte | mask) : (byte)(oldByte & ~mask);

            // Write new byte
            WriteJoinByte(device, writeData, newByte);

            // Return the value written.
            return newByte;
        }

        #endregion
    }
}

File: Framework/Emlid.WindowsIot.HardwarePlus/Buses/I2cExtensions.cpp
#include "pch.h"
#include "I2cExtensions.h"

using namespace Emlid::WindowsIot::HardwarePlus::Buses;

#pragma region Connect

cx::I2cDevice^ I2cExtensions::Connect(int busNumber, int address, cx::I2cBusSpeed speedCX, cx::I2cSharingMode sharingModeCX)
{
	// TODO: Remove argument conversion when C++/WinRT supports components
	winrt::I2cBusSpeed speed = (winrt::I2cBusSpeed)speedCX;
	winrt::I2cSharingMode sharingMode = (winrt::I2cSharingMode)sharingModeCX;

	// C++/WinRT Start -----------------------------------------------------------

	// Validate
	if (busNumber < 0) throw winrt::hresult_invalid_argument(L"busNumber");
	if (address < 0 || address > 0x7f) throw winrt::hresult_invalid_argument(L"address");

	// Query bus information
	auto query = winrt::I2cDevice::GetDeviceSelector();
	auto busInformation = winrt::DeviceInformation::FindAllAsync(query).get();
	if (busInformation.Size() < 1)
		throw winrt::hresult_invalid_argument(L"busNumber");

	// Configure connection
	auto id = busInformation.GetAt(busNumber).Id();
	auto settings = winrt::I2cConnectionSettings(address);
	settings.BusSpeed(speed);
	settings.SharingMode(sharingMode);

	// Connect to device and return (null when failed)
	auto device = winrt::I2cDevice::FromIdAsync(id, settings).get();

	// C++/WinRT End -------------------------------------------------------------

	// Return connected device
	return reinterpret_cast<cx::I2cDevice^>(winrt::detach_abi(device));
}

#pragma endregion


#pragma region Read

byte I2cExtensions::ReadByte(cx::I2cDevice^ device)
{
	// Call overloaded method and return result
	return ReadBytes(device, 1)[0];
}

cx::Array<byte>^ I2cExtensions::ReadBytes(cx::I2cDevice^ device, int size)
{
	// Create buffer
	auto readBuffer = ref new cx::Array<byte>(size);

	// Call extended device method
	device->Read(readBuffer);

	// Return buffer
	return readBuffer;
}

byte I2cExtensions::WriteReadByte(cx::I2cDevice^ device, byte writeData)
{
	// Cast argument to correct type
	auto writeBuffer = ref new cx::Array<byte>(&writeData, 1);

	// Call overloaded method
	return WriteReadBytes(device, writeBuffer, 1)[0];
}

byte I2cExtensions::WriteReadByte(cx::I2cDevice^ device, const cx::Array<byte>^ writeData)
{
	// Call overloaded method
	auto readBuffer = WriteReadBytes(device, writeData, 1);

	// Return first byte in buffer
	return readBuffer[0];
}

cx::Array<byte>^ I2cExtensions::WriteReadBytes(cx::I2cDevice^ device, byte writeData, int size)
{
	// Cast data to correct type
	auto writeBuffer = ref new cx::Array<byte>(&writeData, 1);

	// Call overloaded method
	return WriteReadBytes(device, writeBuffer, size);
}

cx::Array<byte>^ I2cExtensions::WriteReadBytes(cx::I2cDevice^ device, const cx::Array<byte>^ writeData, int size)
{
	// Create buffer
	auto readBuffer = ref new cx::Array<byte>(size);

	// Call extended device method
	device->WriteRead(writeData, readBuffer);

	// Return buffer
	return readBuffer;
}

bool I2cExtensions::WriteReadBit(cx::I2cDevice^ device, byte writeData, byte mask)
{
	// Cast data to correct type
	auto writeBuffer = ref new cx::Array<byte>(&writeData, 1);

	// Call overloaded method and return result
	return WriteReadBit(device, writeBuffer, mask);
}

bool I2cExtensions::WriteReadBit(cx::I2cDevice^ device, const cx::Array<byte>^ writeData, byte mask)
{
	// Read byte
	auto readByte = WriteReadByte(device, writeData);

	// Apply mask and return true when set
	return (readByte & mask) != 0;
}

#pragma endregion

#pragma region Write

void I2cExtensions::WriteByte(cx::I2cDevice^ device, byte writeData)
{
	// Cast data to correct type
	auto writeBuffer = ref new cx::Array<byte>(&writeData, 1);

	// Call extended device function
	device->Write(writeBuffer);
}

/// <summary>
/// Writes one or more data bytes.
/// </summary>
/// <param name="device">Device to use.</param>
/// <param name="data">Data to write.</param>
void I2cExtensions::WriteBytes(cx::I2cDevice^ device, const cx::Array<byte>^ writeData)
{
	// Call extended device function
	device->Write(writeData);
}

/// <summary>
/// Joins two byte values then writes them.
/// </summary>
/// <param name="device">Device to use.</param>
/// <param name="data1">First part of data to write.</param>
/// <param name="data2">Second part of data to write.</param>
void I2cExtensions::WriteJoinByte(cx::I2cDevice^ device, byte writeData1, byte writeData2)
{
	// Create buffer
	auto writeBuffer = ref new cx::Array<byte>(2) { writeData1, writeData2 };

	// Call extended device method
	device->Write(writeBuffer);
}

/// <summary>
/// Joins two byte values then writes them.
/// </summary>
/// <param name="device">Device to use.</param>
/// <param name="data1">First part of data to write.</param>
/// <param name="data2">Second part of data to write.</param>
void I2cExtensions::WriteJoinByte(cx::I2cDevice^ device, const cx::Array<byte>^ writeData1, byte writeData2)
{
	// Cast data to correct type
	auto data2Array = ref new cx::Array<byte>(&writeData2, 1);

	// Call overloaded method
	WriteJoinBytes(device, writeData1, data2Array);
}

/// <summary>
/// Joins two byte values then writes them.
/// </summary>
/// <param name="device">Device to use.</param>
/// <param name="data1">First part of data to write.</param>
/// <param name="data2">Second part of data to write.</param>
void I2cExtensions::WriteJoinBytes(cx::I2cDevice^ device, byte writeData1, const cx::Array<byte>^ writeData2)
{
	// Cast data to correct type
	auto data1Array = ref new cx::Array<byte>(&writeData1, 1);

	// Call overloaded method
	WriteJoinBytes(device, data1Array, writeData2);
}

/// <summary>
/// Joins two byte values then writes them.
/// </summary>
/// <param name="device">Device to use.</param>
/// <param name="data1">First part of data to write.</param>
/// <param name="data2">Second part of data to write.</param>
void I2cExtensions::WriteJoinBytes(cx::I2cDevice^ device, const cx::Array<byte>^ writeData1, const cx::Array<byte>^ writeData2)
{
	// Join buffers
	auto addressLength = writeData1->Length;
	auto dataLength = writeData2->Length;
	auto buffer = ref new cx::Array<byte>(addressLength + dataLength);
	std::memmove(buffer->Data, writeData1->Data, addressLength);
	std::memmove(buffer->Data + addressLength, writeData2->Data, dataLength);

	// Call extended device method
	device->Write(buffer);
}

/// <summary>
/// Sets or clears one or more bits.
/// </summary>
/// <param name="device">Device to use.</param>
/// <param name="writeData">Data to write.</param>
/// <param name="mask">
/// Mask of the bit to set or clear according to value.
/// Supports setting or clearing multiple bits.
/// </param>
/// <param name="value">Value of the bits, i.e. set or clear.</param>
/// <returns>Value written.</returns>
/// <remarks>
/// Commonly used to set register flags.
/// Reads the current byte value, merges the positive or negative bit mask according to value,
/// then writes the modified byte back.
/// </remarks>
byte I2cExtensions::WriteReadWriteBit(cx::I2cDevice^ device, byte writeData, byte mask, bool value)
{
	// Read existing byte
	auto oldByte = WriteReadByte(device, writeData);

	// Merge bit (set or clear bit accordingly)
	auto newByte = value ? (byte)(oldByte | mask) : (byte)(oldByte & ~mask);

	// Write new byte
	WriteJoinByte(device, writeData, newByte);

	// Return the value written
	return newByte;
}

#pragma endregion